api/src/main/java/com/nicknathanjustin/streamercontracts/votes/VoteServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+@Slf4j
+public class VoteServiceImpl implements VoteService{
+
+    @NonNull final VoteModelRepository voteModelRepository;
+
+    @Override
+    public void recordVote(@NonNull final UserModel voter,
+                           @NonNull final ContractModel contractModel,
+                           @NonNull final boolean flaggedCompleted) {
+        if (userCanVoteOnContract(voter, contractModel)) {
+            final VoteModel voteModel = VoteModel.builder()
+                    .voter(voter)
+                    .contract(contractModel)
+                    .viewerFlaggedComplete(flaggedCompleted)
+                    .votedAt(new Timestamp(System.currentTimeMillis()))
+                    .build();
+            voteModelRepository.save(voteModel);
+        }
+    }
+
+    @Override
+    public boolean isVotingComplete(@NonNull final ContractModel contractModel) {
+        validateContract(contractModel);
+
+        final UUID contractId = contractModel.getId();
+        final VoteModel proposerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getProposer().getId());
+        final VoteModel streamerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getStreamer().getId());
+        final Timestamp now = new Timestamp(System.currentTimeMillis());
+        return (proposerVote != null && streamerVote != null) ||
+               (proposerMarkedContractCompleted(contractModel)) ||
+               (streamerMarkedContractFailed(contractModel)) ||
+               (now.after(contractModel.getExpiresAt()));
+    }
+
+    @Override
+    public boolean wasContractCompleted(@NonNull final ContractModel contractModel) {
+        if (!isVotingComplete(contractModel)) {
+            throw new IllegalArgumentException("Attempted to check if contract was completed before voting was finished. ContractId: " + contractModel.getId());
+        }
+
+        final UUID contractId = contractModel.getId();
+        final VoteModel proposerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getProposer().getId());
+        final VoteModel streamerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getStreamer().getId());
+        if (proposerMarkedContractCompleted(contractModel)) {
+            //Proposer has marked contract as completed. Release funds without reading Streamer's vote.
+            return true;
+        } else if (streamerMarkedContractFailed(contractModel)) {
+            //Streamer has marked contract as failed. Return funds without reading Proposers's vote.
+            return false;
+        } else if(proposerVote == null && streamerVote == null) {
+            //Neither proposer or streamer voted. Release funds if contract was accepted. Return funds otherwise.
+            return contractModel.isAccepted();
+        } else if(proposerVote != null) {
+            //Streamer did not vote. Take Proposer's vote as the final outcome
```

Comment:
Wait, this doesn't mean that the streamer did not vote. It just means that the propose did vote. There should be a check for streamerVote == null right?

Reply:
I actually don't think its the same. The above check only verifies that they're both null. If proposerVote != null, streamerVote can be null or !null. 

Good that its's switched.

Reply:
Ahhh, my b. Yup, good catch